### PR TITLE
Add Connection Source

### PR DESCRIPTION
**Pull request by drewaogle**
adds functions to expose formatted connection information

### CODE REVIEW

File: src/comm/TCPSocket.cc [hunk 1]
```diff
@@ -21,8 +21,10 @@ ENABLE_WARNING(effc++)
 #include "comm/Variables.h"
 
 using namespace comm;
-
-TCPSocket::TCPSocket(int socket_fd) : _socket_fd(socket_fd) {}
+TCPSocket::TCPSocket(int socket_fd, const sockaddr_in& address)
+    : _socket_fd(socket_fd), _source_family(AF_INET), _source(address)
```
Comment (JoshStoddard) [inline src/comm/TCPSocket.cc:25]: nit: If `sockaddr_in` is moveable, prefer pass-by-value and move in cases where the method makes a copy:
```suggestion
TCPSocket::TCPSocket(int socket_fd, sockaddr_in address)
    : _socket_fd(socket_fd), _source_family(AF_INET), _source(std::move(address))
```

File: src/comm/TCPSocket.cc [hunk 2]
```diff
@@ -66,7 +68,7 @@ std::unique_ptr< TCPSocket > TCPSocket::accept(const std::unique_ptr< TCPSocket
     }
     // MAGICK can be done here
 
-    return std::unique_ptr< TCPSocket >(new TCPSocket(connected_socket));
+    return std::unique_ptr< TCPSocket >(new TCPSocket(connected_socket, clnt_addr));
```
Comment (JoshStoddard) [inline src/comm/TCPSocket.cc:71]:
```suggestion
    return std::unique_ptr< TCPSocket >(new TCPSocket(connected_socket, std::move(clnt_addr)));
```

File: src/comm/TCPSocket.cc [hunk 3]
```diff
@@ -104,12 +106,14 @@ bool TCPSocket::connect(const Address& address)
 std::unique_ptr< TCPSocket > TCPSocket::create()
 {
     int tcp_socket = ::socket(AF_INET, SOCK_STREAM, 0);
+    sockaddr_in source;
+    memset(&source, 0, sizeof(_source));
```
Comment (JoshStoddard) [inline src/comm/TCPSocket.cc:110]: Yikes, this `memset` is not a good code smell.  Does `sockaddr_in` not initialize itself properly?

However we initialize `sockaddr_in`, I feel like it should happen in `TCPSocket` ctor instead of being the caller's responsibility.

File: include/comm/Connection.h [hunk 4]
```diff
@@ -68,6 +74,14 @@ class Connection
     uint32_t _max_buffer_size{};
 
     ConnMetrics* _metrics{nullptr};
+
+    std::string _source{};
+    std::string _encryption{};
+    short _source_family;  // from socket(2)
```
Comment (JoshStoddard) [inline include/comm/Connection.h:80]: Why duplicate these values here?  Can't we just get them from the socket?

File: include/comm/Connection.h [hunk 5]
```diff
@@ -68,6 +74,14 @@ class Connection
...
+
+    friend class ConnectionTester;
+
+    static std::shared_ptr< Connection > testingConnection();
```
Comment (JoshStoddard) [inline include/comm/Connection.h:84]: Why is this necessary?

File: src/comm/Connection.cc [hunk 6]
```diff
@@ -125,8 +127,39 @@ void Connection::set_max_buffer_size(uint32_t max_buffer_size)
...
+   public:
+    TestConnection() : Connection() { _source = "testing"; }
+    size_t read(uint8_t* /*buffer*/, size_t /*length*/) override { return 0; }
+    size_t write(const uint8_t* /*buffer*/, size_t /*length*/) override { return 0; }
+};
+
+std::shared_ptr< Connection > Connection::testingConnection()
+{
+    return std::make_shared< TestConnection >();
+}
```
Comment (JoshStoddard) [inline src/comm/Connection.cc:165]: Does this really need to be here?

**Review by JoshStoddard — CHANGES_REQUESTED** [review]
I think I see what you're doing here, but it needs to be cleaned up.

Also, can you link to the sister PR in athena?

**Comment by drewaogle** [comment]
> I think I see what you're doing here, but it needs to be cleaned up.
> 
> Also, can you link to the sister PR in athena?

certainly - I've made it a draft - https://github.com/aperture-data/athena/pull/691

File: src/comm/Connection.cc (hunk 6 above)
Comment (drewaogle) [reply src/comm/Connection.cc:165]: it doesn't need to be HERE, but what needs to happen is in athena the test framework needs to create a connection to feed into the testing framework, so this was the first logical place to put it.

File: include/comm/Connection.h [hunk 7]
```diff
@@ -60,6 +61,11 @@ class Connection
     void set_max_buffer_size(uint32_t max_buffer_size);
     bool check_message_size(uint32_t size);
 
+    std::string source_family_name(short source_family);
```
[the remark this answers is not in the document]
Comment (drewaogle) [reply include/comm/Connection.h:64]: uncalled in this PR - used in athena.

File: src/comm/TCPSocket.cc (hunk 1 above)
Comment (drewaogle) [reply src/comm/TCPSocket.cc:25]: I'm a fan of move, but know it can be tricky. I'll check the conditions.

File: src/comm/TCPSocket.cc (hunk 3 above)
Comment (drewaogle) [reply src/comm/TCPSocket.cc:110]: sockaddr_in is pure C, so my assumption was that you treat it like a C struct.

File: include/comm/Connection.h (hunk 4 above)
Comment (drewaogle) [reply include/comm/Connection.h:80]: the values aren't duplicate here, this is where they are rendered into a uniform format. So my thinking is that if we were to make a unix socket connection at some point, it would be behind this abstraction, and this would be the point it would be unified for the metrics user of this data.

Comment (JoshStoddard) [reply include/comm/Connection.h:80]: I'm not sure I follow. It looks to me like these members are simply caching values returned by the socket. My question is why cache these values inside the base class?  Wouldn't a virtual interface give you the same abstraction with a smaller footprint per connection and potentially fewer string copies?  Eg.:
```cpp
class Connection
{
   public:
    // ...

    virtual std::string get_source() const = 0;
    virtual short get_source_family() const = 0;
    virtual std::string get_encryption() const = 0;
};


class TCPConnection : public Connection
{
   public:
    // ...

    std::string get_source() const override { return _tcp_socket->print_source(); }
    short get_source_family() const override { return _tcp_socket->source_family(); }
    std::string get_encryption() const { return "none"; }
};


class TLSConnection : public Connection { /* as above */ };


// possibly in the future
class UNIXConnection : public Connection
{
   public:
    // ...

    std::string get_source() const override { /*...*/ }
    short get_source_family() const override { /*...*/ }
    std::string get_encryption() const { /*...*/ }
};
```

File: src/comm/TCPSocket.cc (hunk 3 above)
Comment (drewaogle) [reply src/comm/TCPSocket.cc:110]: The way `TCPSocket` currently works is that there is only 1 constructor - it's internal and so it always is going to pass a valid `_source`. The design of the class is essentially a base class which is kind of a factory ( valid objects are created by accept or create ) and create is done the way it is because of how `ConnClient` utilizes it and generating an actually valid socket only after `connect` is called.

File: src/comm/TCPSocket.cc (hunk 1 above)
Comment (drewaogle) [reply src/comm/TCPSocket.cc:25]: since sockaddr_in is a struct inside a extern "C", it can't have an automatic move constructor generated, so a copy is the best we can ask for, unless I'm reading things wrong.

File: src/comm/TCPSocket.cc (hunk 3 above)
Comment (JoshStoddard) [reply src/comm/TCPSocket.cc:110]: >The way `TCPSocket` currently works is that there is only 1 constructor

Why is it important for `TCPSocket` to have a single ctor? Por que no los dos? 

It seems to me if we're adding a new member to `TCPSocket`, shouldn't it be that class's responsibility to initialize it correctly? That strikes me as an implementation detail of `TCPSocket` that this method shouldn't have to care about.

Comment (drewaogle) [reply src/comm/TCPSocket.cc:110]: Right - and the way it is initialized properly is by requiring it to be passed in, just like the fd is required.

File: include/comm/Connection.h (hunk 4 above)
Comment (drewaogle) [reply include/comm/Connection.h]: modified to be virtual interface

**Review by JoshStoddard — APPROVED** [review]
Thanks

**Comment by luisremis** [comment]
mmm this should be squashed merge to remove all the ugly commits. In general, we always squash merge. I will revert this and merge it again.
